parse.y

```diff
@@ -7564,35 +7623,98 @@ parser_has_token(struct parser_params *p)
 {
-    VALUE ary;
-    int token_id;
+    rb_encoding *enc = p->enc;
+    const char *ptr = str->ptr;
+    const char *pend = ptr + str->len;
+    const char *prev = ptr;
+    char charbuf[5] = {'\\', 'x', 0, 0, 0};
+    rb_parser_string_t * result = rb_parser_string_new(p, 0, 0);
+    int asciicompat = rb_enc_asciicompat(enc);
+
+    while (ptr < pend) {
+        unsigned int c;
+        const char *cc;
+        int n = rb_enc_precise_mbclen(ptr, pend, enc);
+        if (!MBCLEN_CHARFOUND_P(n)) {
+            if (ptr > prev) rb_parser_str_buf_cat(p, result, prev, ptr - prev);
+            n = rb_enc_mbminlen(enc);
+            if (pend < ptr + n)
+                n = (int)(pend - ptr);
+            while (n--) {
+                c = *ptr & 0xf0 >> 4;
+                charbuf[2] = (c < 10) ? '0' + c : 'A' + c - 10;
+                c = *ptr & 0x0f;
+                charbuf[3] = (c < 10) ? '0' + c : 'A' + c - 10;
+                rb_parser_str_buf_cat(p, result, charbuf, 4);
+                prev = ++ptr;
+            }
+            continue;
+        }
+        n = MBCLEN_CHARFOUND_LEN(n);
+        c = rb_enc_mbc_to_codepoint(ptr, pend, enc);
+        ptr += n;
+        cc = escaped_char(c);
+        if (cc) {
+            if (ptr - n > prev) rb_parser_str_buf_cat(p, result, prev, ptr - n - prev);
+            rb_parser_str_buf_cat(p, result, cc, strlen(cc));
+            prev = ptr;
+        }
+        else if (asciicompat && rb_enc_isascii(c, enc) && ISPRINT(c)) {
+        }
+        else {
+            if (ptr - n > prev) {
+                rb_parser_str_buf_cat(p, result, prev, ptr - n - prev);
+                prev = ptr - n;
+            }
+            rb_parser_str_buf_cat(p, result, prev, ptr - prev);
+            prev = ptr;
+        }
+    }
+    if (ptr > prev) rb_parser_str_buf_cat(p, result, prev, ptr - prev);
+
+    return result;
+}
 
-    ary = rb_ary_new2(4);
-    token_id = p->token_id;
-    rb_ary_push(ary, INT2FIX(token_id));
-    rb_ary_push(ary, ID2SYM(parser_token2id(p, t)));
-    rb_ary_push(ary, str);
-    rb_ary_push(ary, code_loc_to_ary(p, p->yylloc));
-    rb_obj_freeze(ary);
-    rb_ary_push(p->tokens, ary);
+static void
+parser_append_tokens(struct parser_params *p, rb_parser_string_t *str, enum yytokentype t, int line)
+{
+    rb_parser_ast_token_t *token = xcalloc(1, sizeof(rb_parser_ast_token_t));
+    token->id = p->token_id;
+    token->type_name = parser_token2char(p, t);
+    token->str = str;
+    token->loc.beg_pos = p->yylloc->beg_pos;
+    token->loc.end_pos = p->yylloc->end_pos;
+    rb_parser_ary_push(p, p->tokens, (VALUE)token);
     p->token_id++;
 
     if (p->debug) {
-        rb_parser_printf(p, "Append tokens (line: %d) %"PRIsVALUE"\n", line, ary);
+        rb_parser_printf(p, "Append tokens (line: %d) [%d, :%s, \"%s\", [%d, %d, %d, %d]]\n",
+                            line, token->id, token->type_name, rb_parser_str_escape(p, token->str)->ptr,
```

Comment:
I guess `rb_parser_string_t` allocated in `rb_parser_str_escape` is not freed.

Reply:
Thanks. Fixed in https://github.com/hasumikin/ruby/commit/edd7c6d563523d6fe99efaeda800ad69d60948c7